Jetpack_Gutenberg: Simplify

ockham opened the pull request:
Refactor `Jetpack_Gutenberg` to get rid of some of its quirks, and make it easier to reason about. Among other things, this allows us to remove the Mailchimp workaround (https://github.com/Automattic/jetpack/pull/10597/files#r245018740) as part of this PR (per 293b373109f49231db13f1c726298d7655115c99).

Some basic ideas:

- Decouple registration, and storing of availability information. Only when a block isn't being registered do we set the latter.
- No need for a duplicated block registry -- simply use Gutenberg's.
- Fewer getters/setters in `Jetpack_Gutenberg`, thus fewer (potentially broken) states. The idea is that `Jetpack_Gutenberg` initializes a whitelist of potentially allowed blocks and plugins once, and upon (attempted) registration of individual blocks and plugins, checks if they're on that list. This is probably best reflected by the changes to the unit test.
- Remove layers of indirection, especially wrt filters. Instead of passing a callback when setting availability in more complex scenarios, we simply move block registration and availability setting to an appropriate filter _inside_ the corresponding module. This should make the order of actions and flow of filters much easier to reason about.
- Removal of asymmetry (there were some methods for plugins that didn't actually do anything).
- Filter child blocks from availability tracking so we don't confuse them with other unwhitelisted blocks.

#### Changes proposed in this Pull Request:

- [x] Introduce and use set_extension_unavailability_reason() (to decouple block registration and availability)
- [x] Update endpoint to directly fetch registration information from GB block registry
- [x] Remove obsolete stuff from `Jetpack_Gutenberg`
- [x] Rebase, remove registration workarounds from #10597
- [x] Fix (and update) tests
- [x] Update comments

#### Testing instructions:

Verify that block availability works as before. Verify that the following extensions's availability changes when doing the following:
- Publicize:
  - Toggle the Publicize module, verify that the Publicize panel only shows when the module is on
  - With a user with 'contributor' role, write a new post, and verify that the Publicize panel doesn't show in the Jetpack sidebar. (If you check `window.Jetpack_Editor_Initial_State`, it should say `unavailable_reason: "unauthorized"` for `publicize`.)
- Simple Payments: Try a free vs a premium plan
- Contact Form:
  - Start writing a post, and insert a Contact Form Block
  - Modify some of the Contact Form's field labels (e.g. change 'Name' to 'Your Name')
  - Publish the post, and view it in the frontend.
  - Verify that your modifications actually show up.

Change the `JETPACK_BETA_BLOCKS` constant and verify that depending on its value, beta blocks are either available or not!

#### Proposed changelog entry for your changes:

* Simplify the `Jetpack_Gutenberg` class

ockham committed:
Jetpack Gutenberg: Kill `$default_blocks` and `$default_plugins` with :fire: (#11075)

ockham committed:
Introduce and use set_extension_unavailability_reason()

ockham committed:
Getting there

ockham committed:
Remove deprecated load_blocks() method

ockham committed:
Remove deprecated get_block_availability() method

ockham committed:
Remove redundant ::load() invocation

ockham committed:
Fall back to 'missing_module'

ockham committed:
Add an underscore in 'not_whitelisted'

ockham committed:
Singleton pattern

ockham committed:
Remove more obsolete stuff

ockham committed:
Take plugins into account

ockham committed:
De-dupe set_jetpack_extension_availability()

ockham committed:
Remove obsolete availability arg from jetpack_register_plugin()

ockham committed:
Publicize: Decouple Plugin Registration and Availability

ockham committed:
Simplify second set_jetpack_extension_availability() argument

ockham committed:
Add comments

ockham committed:
Remove obsolete beta qarg functionality

ockham committed:
Remove Mailchimp block registration workaround

ockham committed:
Include unwhitelisted extensions

ockham committed:
Exclude child blocks

ockham committed:
Register child blocks if their parent is whitelisted

ockham committed:
Remove now-obsolete stuff

ockham committed:
Remove jetpack_set_available_blocks()'s argument

ockham committed:
Rename jetpack_set_available_blocks() to get_jetpack_gutenberg_extensions_whitelist

ockham committed:
Re-introduce filter

ockham committed:
Add uninit() method

ockham committed:
Ditch singleton pattern again

ockham committed:
Make availability check stabler

ockham committed:
Fix uninit(), rename to reset()

ockham committed:
Update tests

ockham committed:
Rename set_jetpack_extension_availability to set_extension_unavailability_reason

ockham committed:
Remove deprecated should_load_blocks() method

ockham committed:
Update two comments

enejb commented on class.jetpack-gutenberg.php:
```diff
@@ -10,143 +10,152 @@
+ *
+ * @return void
+ */
+function set_extension_unavailability_reason( $slug, $reason ) {
```
lets prefix this with `jetpack_`

ockham replied:
Done in e63096d.

enejb commented on class.jetpack-gutenberg.php:
```diff
@@ -10,143 +10,152 @@
+	 *
+	 * @return boolean True if $a and $b share at least one item
 	 */
-	public static function register( $slug, $args, $availability ) {
```
lets properly deprecate this method.

ockham replied:
579269f

enejb commented on class.jetpack-gutenberg.php:
```diff
@@ -155,109 +164,31 @@ static function load( $response = null, $handler = null, $request = null ) {
-
-	static function is_registered( $slug ) {
-		return isset( self::$registered[ $slug ] );
+		self::$extensions = apply_filters( 'jetpack_set_available_blocks', self::get_jetpack_gutenberg_extensions_whitelist() );
```
Are we still calling this filter at the same runtime as before?
We it used to be called at action `init` priority 90. This would make it more usable in plugins.

ockham replied:
tl;dr: It's still possible to add the filter prior to initialization.

So the previous behavior was:

- Define an array of [`$default_blocks`](https://github.com/Automattic/jetpack/blob/91d89bc2033bf42fd6993ac712878866f46d4a5b/class.jetpack-gutenberg.php#L48-L66).
- In `init()`, [apply](https://github.com/Automattic/jetpack/blob/91d89bc2033bf42fd6993ac712878866f46d4a5b/class.jetpack-gutenberg.php#L182) the `jetpack_set_available_blocks` filter to those.
- Define a static method _also_ called [`jetpack_set_available_blocks`](https://github.com/Automattic/jetpack/blob/91d89bc2033bf42fd6993ac712878866f46d4a5b/class.jetpack-gutenberg.php#L301-L323).
- That method is added to the filter of the same name _outside_  of the class, in `class.jetpack.php`: https://github.com/Automattic/jetpack/blob/91d89bc2033bf42fd6993ac712878866f46d4a5b/class.jetpack.php#L543

This has been very confusing to me, and I'd say that it arguably breaks encapsulation: if we're both applying the filter _and_ providing a default method to use it with that even bears the same name, it's very surprising to see that filter added outside the class.

Furthermore, `$default_blocks` has long been confusing people, see #11075.

So removing `$default_blocks` gives us the opportunity to use the return value of the `jetpack_set_available_blocks` _method_ (which I'm renaming, to tell it apart from the filter) as the 'default' in `apply_filters`, while still giving outside stakeholders the chance to add their own filters on top of that. This is e.g. still used in the unit test.

There are conceptual reasons why I believe we shouldn't modify the whitelist of allowed blocks after initialization. In OOP parlance, this is our invariant: If we change it after the fact, what does that mean for blocks that were registered before that change? It's easy to create a mess of badly defined states here. I'll say that ideally, I think this class should become even more encapsulated -- i.e. turn a lot of `static` stuff into regular instance methods and variables, turn the `init` into a constructor. (We would still leave it to consumers to invoke that constructor by creating a new object, and could pass data like the whitelist as an arg to the constructor.) This is the kind of OOP tooling that allows to preserve those invariants (and hence keep state sane). I didn't make that change because it would've inflated the size of this PR even more.

ockham committed:
Rename set_extension_unavailability_reason to jetpack_set_extension_unavailability_reason

sirreal commented on class.jetpack-gutenberg.php:
```diff
@@ -10,143 +10,152 @@
-		$sanitized_slug = sanitize_title_with_dashes( $slug );
-		self::$registered[ $sanitized_slug ] = array( 'args' => $args, 'availability' => $availability );
+	protected static function share_items( $a, $b ) {
+		return count( array_intersect( $a, $b ) ) > 0;
```
Equivalent and more descriptive?

```suggestion
		return ! empty( array_intersect( $a, $b ) );
```

ockham replied:
I had it like that, but then I was reminded that prior to PHP 5.5, `empty()` would only work for variables (but not for expressions) and would apparently yield a parse error for anything else.

[SO thread](https://stackoverflow.com/questions/523796/checking-to-see-if-one-arrays-elements-are-in-another-array-in-php) about checking if intersections of arrays are empty where this is coincidentally brought up.

sirreal reviewed (commented):
Left a nit.

There's a lot going on here and I don't feel that I have enough context in the ongoing discussion around block availability to give an authoritative review. The code changes I've looked over make sense and if this tests well, I don't see any blockers.

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -155,109 +164,31 @@ static function load( $response = null, $handler = null, $request = null ) {
-		 *
-		 * @param array
-		 */
-		self::$plugins = apply_filters( 'jetpack_set_available_plugins', self::$default_plugins );
```
Seems like we're removing this filter completely. Instead, it should be deprecated instead of removed - it was out in 1 or more versions, so someone could be relying on it.

ockham replied:
Renamed the still-existing `jetpack_set_available_plugins` to `jetpack_set_available_extensions` (3ab63a5c98ed2ee881e72f6bcc804650cbd10c6d), and brought both `jetpack_set_available_blocks` and `jetpack_set_available_plugins` back, deprecating both (9a1abea6f901698610823ba07e341a7013da3bdf).

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -10,143 +10,152 @@
+		}
 
-	static function load( $response = null, $handler = null, $request = null ) {
-		$is_availability_endpoint_beta = ! is_null( $request ) && $request->get_param( 'beta' ) && wp_endswith( $request->get_route(), 'gutenberg/available-extensions' );
```
We seem to unconditionally be removing this logic. Is that intentional? Are we sure we're not relying on it somewhere?

enejb replied:
The whole idea behind the flag was to add the ability for any API client to be able to include the beta blocks weather the user is proxied or not. Removing the argument make that impossible.

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -155,109 +164,31 @@ static function load( $response = null, $handler = null, $request = null ) {
-		 *
-		 * @param array
-		 */
-		$blocks = apply_filters( 'jetpack_blocks_to_register', self::$blocks );
```
We're removing this one, should we find a way to deprecate it?

ockham replied:
See https://github.com/Automattic/jetpack/pull/11091#discussion_r246918397

ockham replied:
That's the Mailchimp hack :slightly_frowning_face: There's not really any equivalent functionality left per this PR, since it's something I wanted to explicitly get rid of.

I think I'll have to risk breaking backward compat here. I think chances of anyone using this outside A8c are pretty minimal.

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -155,109 +164,31 @@ static function load( $response = null, $handler = null, $request = null ) {
+	 *
+	 * @return void
+	 */
+	public static function reset() {
```
This is what happens when we use classes with only static methods.

ockham replied:
Indeed. I hated myself a tiny bit when writing this.

ockham replied:
See https://github.com/Automattic/jetpack/pull/11091#discussion_r246811623 which includes some thoughts on turning this class into an actually instantiable one.

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -299,46 +230,52 @@ public static function get_preset( $preset ) {
+				'available' => $is_available,
+			);
+
+			if ( ! $is_available ) {
```
Maybe a ternary could help shorten this code:

```
if ( ! $is_available ) {
	$reason = isset( self::$availability[ $extension ] ) ? self::$availability[ $extension ] : 'missing_module';
	$available_extensions[ $extension ][ 'unavailable_reason' ] = $reason;
}
```

tyxla commented on modules/publicize/publicize.php:
```diff
@@ -780,17 +780,15 @@ function add_post_type_support() {
 	function register_gutenberg_extension() {
-		jetpack_register_plugin( 'publicize', array( 'callback' => array( $this, 'get_extension_availability' ) ) );
-	}
+		// TODO: The `gutenberg/available-extensions` endpoint currently doesn't accept a post ID,
```
We have to fix this - it currently introduces a regression.

ockham replied:
I don't think it does, TBH. Keep in mind that in case of Publicize, this is only used for the availability endpoint, so this will only be evaluated when that endpoint is it. (Gutenberg itself doesn't even have a concept of server-side plugin registration -- while there's `register_block_type`, there's no `register_plugin`).

However, that [endpoint](https://github.com/Automattic/jetpack/blob/91d89bc2033bf42fd6993ac712878866f46d4a5b/_inc/lib/core-api/wpcom-endpoints/gutenberg-available-extensions.php) doesn't currently support a `post` argument (and didn't before -- which is my fault, since I think Miguel was suggesting to add one). So [`_GET['post']`](https://github.com/Automattic/jetpack/pull/11091/files#diff-2f5f58835a08193f26adccabf4ae5ef3L787) was always empty, meaning that `current_user_can_access_publicize_data()` was always only called generically (checking the user's `publish_posts` capability in general, rather than for a given post ID specifically.

tyxla commented on modules/publicize/publicize.php:
```diff
@@ -119,7 +119,7 @@ function __construct() {
 		add_action( 'init', array( $this, 'add_post_type_support' ) );
 		add_action( 'init', array( $this, 'register_post_meta' ), 20 );
-		add_action( 'init', array( $this, 'register_gutenberg_extension' ), 30 );
+		add_action( 'rest_api_init', array( $this, 'register_gutenberg_extension' ), 30 );
```
Do we need this `30` priority?

ockham replied:
(Since I've reverted `rest_api_init` to the previous `init`, I guess I'm also leaving the `30` untouched so the line is entirely unchanged.)

tyxla commented on modules/publicize/publicize.php:
```diff
@@ -119,7 +119,7 @@ function __construct() {
 		add_action( 'init', array( $this, 'add_post_type_support' ) );
 		add_action( 'init', array( $this, 'register_post_meta' ), 20 );
-		add_action( 'init', array( $this, 'register_gutenberg_extension' ), 30 );
+		add_action( 'rest_api_init', array( $this, 'register_gutenberg_extension' ), 30 );
```
This way we won't have the extension registered for non-REST-API requests. Are we sure that's what we want, and that this won't be useful outside of the REST API context?

ockham replied:
I'll revert back to `init`, since I've just verified that that still works. (I think I made the change at some point into the process when I still had some of `Jetpack_Gutenberg`'s own initialization depend too much on that hook and it didn't work.)

Some more general notes on the question:

I've also found myself wondering about this while working on this PR (and looking for appropriate actions to hook into). For most cases, we currently really only seem to need block registration and availability status for the `/gutenberg/available-extensions` endpoint (and not much otherwise). There's at least one counter-example I know of: If the contact form's child blocks aren't registered on the server side, they will show their default content on the frontend instead of whatever changes the user has made to them. (This was the bug caused by #11075 that required me to revert it.)

If we also start using Gutenberg's own [mechanism](https://github.com/WordPress/gutenberg/blob/95edac1e42cb10ed7da9f406696cdc85d6e476d5/lib/client-assets.php#L826-L833) to enqueue frontend-side block assets (which I've discovered only recently) instead of `Jetpack_Gutenberg`s (which I think would make total sense), server/side block registration will also be relevant for that.

ockham replied:
I think the case I was most worried about here (and that prompted me to change the action to hook into) was that previously, we were using block registration via a callback here (which was one of `Jetpack_Gutenberg`'s previous workarounds for problems with initialization order). Specifically, we want to set Publicize availability to `false` (because `unauthorized`) for users that don't have sufficient capabilities. 

I just changed back the hook to `init` locally and tested with a user with `contributor` role, and Publicize is indeed not showing up (and `unavailable_reason` set to `unauthorized`). I'll commit the change then :smile:

ockham replied:
dc2d006

tyxla commented on tests/php/sync/test_class.jetpack-sync-callables.php:
```diff
@@ -87,7 +87,7 @@ public function test_sync_callable_whitelist() {
 			'roles'                            => Jetpack_Sync_Functions::roles(),
 			'timezone'                         => Jetpack_Sync_Functions::get_timezone(),
-			'available_jetpack_blocks'         => Jetpack_Gutenberg::get_block_availability(),
+			'available_jetpack_blocks'         => Jetpack_Gutenberg::get_availability(),
```
Could `available_jetpack_blocks ` also be renamed to accommodate for the fact that this array contains plugins too?

ockham replied:
I didn't dare to change this because it's a the name of a synced value, so I was worried it might show up wrong for existing sites that are already using this name. But I have no idea if that is really a problem.

Maybe a @Automattic/jetpack-crew member can advise if it's fine to simply change this name? :smile:

dereksmart replied:
As discussed in Slack, not so simple. It would need to be accounted for in the wpcom-side whitelist, and then leave us with another question of whether or not we keep the old key to account for older Jetpack versions. There might be other things this affects that I am not sure of, such as Activity Log (maybe)?

I'd say it's not worth the effort to rename, and instead use our old friend: inline comments.

ockham replied:
This is the test file, I'm adding the comment to `sync/class.jetpack-sync-defaults.php` instead.

ockham replied:
4f027dbe8

tyxla reviewed (commented):
I like where this is heading 👍 Nice cleanup!

I've left some comments and questions.

I also see some good reasons for a larger refactor - currently `Jetpack_Gutenberg` is a pack of functionality and has multiple responsibilities and concerns. To add to that, every method is static, so it doesn't help us much. I'd love to see different classes with different responsibilities - one for registry, one for availability, one for assets, and maybe one for loading/wrapping. Not in this PR of course - it's already huge 😉 

I haven't tested this - let's make sure to test it thoroughly before we 🚢 :

* In all environments (WP.com wp-admin, Calypso, Jetpack)
* With all blocks and extensions.
* With various conditions (with/without paid plan, with/without modules enabled) where applicable.

ockham committed:
Register publicize at init time again

ockham committed:
/available-extensions endpoint: Remove ?beta query arg

We would need to retain some unnecessary complexity to keep this, and it's kind of redundant now that we show beta blocks when proxied.

ockham committed:
Shorten code a bit. Props @tyxla

ockham committed:
Rename jetpack_set_available_blocks filter to jetpack_set_available_extensions

dereksmart commented on class.jetpack-gutenberg.php:
```diff
@@ -299,46 +230,49 @@ public static function get_preset( $preset ) {
 	public static function get_availability() {
-		if ( ! self::should_load() ) {
-			return array();
+		$available_extensions = [];
```
This array syntax is failing PHP 5.2 tests

ockham replied:
0222a32

ockham committed:
Add back jetpack_set_available_blocks and jetpack_set_available_plugins filters, deprecate

ockham committed:
Array syntax fix for PHP 5.2 compat

ockham committed:
Inline comment about available_jetpack_blocks sync var

ockham committed:
Add back, and deprecate register()

ockham committed:
Try to make tests work in Travis

ockham committed:
Try to make tests work in Travis

ockham committed:
Revert "Try to make tests work in Travis"

This reverts commit f6deee9554549c5189cee2e4b7706cee36fbfbbd.

ockham committed:
Sync tests: Tear down properly

ockham committed:
Typos

ockham committed:
Fix tests, again

ockham committed:
Remove workaround

ockham committed:
Too much teardown

ockham commented on _inc/lib/core-api/wpcom-endpoints/gutenberg-available-extensions.php:
```diff
@@ -28,12 +28,6 @@ public function register_routes() {
 				'methods'             => WP_REST_Server::READABLE,
 				'callback'            => array( 'Jetpack_Gutenberg', 'get_availability' ),
 				'permission_callback' => array( $this, 'get_items_permission_check' ),
-				'args'                => array(
```
See https://github.com/Automattic/jetpack/pull/11091#discussion_r246802557 and my reply to it :)

ockham commented on modules/simple-payments/simple-payments.php:
```diff
@@ -57,14 +58,14 @@ public function init_hook_action() {
-		return $this->is_enabled_jetpack_simple_payments()
-		       ? array( 'available' => true )
-		       : array( 'available' => false, 'unavailable_reason' => 'missing_plan' );
+	function register_gutenberg_block() {
```
I thought `init` was too early for plans information to be available (which we need for the `$this->is_enabled_jetpack_simple_payments()` check), but maybe I was mistaken :man_shrugging:

ockham replied:
75a0b0196

ockham committed:
Register Simple Payments upon init instead of rest_api_init

brbrr reviewed (dismissed):
Seems like Contact From block is get broken. It fails to render in block-editor with the following error:
`This block has encountered an error and cannot be previewed.`

error in DevTools:
> TypeError: Cannot read property ‘attributes’ of undefined

Publicize & Payment button block works like a charm 👌

ockham commented:
Thanks for testing, @brbrr. As discussed in Slack, I can't seem to reproduce -- the Contact Form block seems to work just fine for me, both in my local Docker testing environment, and jurassic.ninja. I wonder what's up :thinking:

jeherve committed:
phpcs fixes

jeherve commented:
I'm running into the same error as @brbrr on my setup (5.0.3, no Gutenberg). This happens whether I use the beta blocks or not.

jeherve reviewed (approved):
After building a fresh version of the blocks from Calypso `master`, everything works well! 🚢
